torch/distributed/_shard/sharded_tensor/_ops/matrix_ops.py

```diff
@@ -1,76 +1,328 @@
-        input.size(),  # type: ignore[arg-type]
-        process_group=pg,
-    )
+    dim = dim if dim is not None else 1  # If no dim specified, softmax use 1 as dim.
+    if dim == st.sharding_spec().dim:
+        raise NotImplementedError(
+            "Only support performing softmax on non-sharding dim now."
+        )
+
+
+_register_sharded_op_on_local_tensor(
+    torch.nn.functional.softmax,
+    extra_check=sharded_softmax_check,
+)
+
+
+def sharded_masked_fill_check(*args, **kwargs):
+    """
+    Perform extra checks for the sharded_masked_fill op.
+    Ensure the mask size is compatible with the size of
+    the sharded tensor.
+
+    Args: same as ``torch.Tensor.masked_fill``.
+
+    Return: None
+    """
+    st = args[0]
+    mask = args[1]
+    if st.dim() < mask.dim():
+        raise ValueError(
+            "mask dim must not greater than the dim of the sharded tensor."
+        )
+    for idx in range(-1, -mask.dim() - 1, -1):
+        if mask.size(idx) != st.size(idx):
+            raise ValueError(
+                f"The size of mask {mask.dim() + idx} must match the size of "
+                f"sharded tensor {st.dim() + idx} at non-singleton dimension {mask.dim() + idx}"
+            )
+
 
-@sharded_op_impl(torch.nn.functional.dropout)
-def sharded_dropout(types, args=(), kwargs=None, pg=None):
+def sharded_masked_fill(args, kwargs, pg):
     """
-    Handles ``__torch_function__`` dispatch for the Dropout op.
+    Handles ``__torch_function__`` dispatch for the sharded_masked_fill op.
+    We first narrow down the mask to the size of local tensor if the mask
+    contains the sharding dim and then apply the mask to the local tensor.
+
+    Args: same as ``torch.Tensor.masked_fill``.
+
+    Return:
+        local_tensor (Tensor): New local tensor to build the sharded tensor.
+        sharding_spec (:class:`torch.distributed._shard.sharding_spec.ShardingSpec`):
+            sharding spec of the new sharded tensor.
+        new_st_size (torch.Size): Size of the new sharded tensor.
     """
-    input = args[0]
-    p = kwargs.get("p")
+    st = args[0]
+    mask = args[1]
+    value = args[2]
+    current_rank = dist.get_rank(pg)  # type: ignore[attr-defined]
+    sharding_dim = st.sharding_spec().dim  # type: ignore[attr-defined]
+    narrow_idx = None
+    for idx in range(-1, -mask.dim() - 1, -1):
+        if st.dim() + idx == sharding_dim:
+            narrow_idx = idx
+    if narrow_idx is not None:
+        rank_idx = None
+        for idx, placement in enumerate(st._sharding_spec.placements):  # type: ignore[attr-defined]
+            if placement.rank() == current_rank:  # type: ignore[index]
+                rank_idx = idx  # type: ignore[attr-defined]
+        shard_metadata = st.metadata().shards_metadata[rank_idx]  # type: ignore[index]
+        mask = mask.narrow(
+            narrow_idx,
+            shard_metadata.shard_offsets[sharding_dim],
+            shard_metadata.shard_sizes[sharding_dim],
+        )
```

Comment:
So this is assuming the mask is a ReplicatedTensor which is the same on all ranks and the value is a regular torch.Tensor which is different on each rank?

In the local version of this op, the original tensor, mask and value are of same size. But here it seems a bit different? Maybe mask needs to be a ReplicatedTensor and value needs to be a ShardedTensor?

In our particular Transformer use case, how are these tensors set up?

Reply:
They are same across all ranks. We just need to narrow them to the part which is stored on the current rank.